# eda pc


Here will be the basic eda of the partitioned data 




In [1]:
import org.apache.spark.sql.types.LongType
var df = spark.read.load("../Projects/data/partitioned_hdfs")
  .withColumn("id",col("id").cast(LongType))
  .dropDuplicates("id")
  
  

In [2]:
df.show()

+--------------------+-------+--------+----------+
|              claims|     id|group_id|      date|
+--------------------+-------+--------+----------+
|The floor cleanin...|3930281|    A47L|1976-01-06|
|The process of cl...|3930734|    G01D|1976-01-06|
|A membrane formin...|3930886|    B01D|1976-01-06|
|Method according ...|3930897|    H01F|1976-01-06|
|The method of cla...|3930951|    C12N|1976-01-06|
|A method as in cl...|3931049|    C22C|1976-01-06|
|to 95 weight perc...|3931102|    C08K|1976-01-06|
|A compound as cla...|3931222|    C07D|1976-01-06|
|A compound as def...|3931227|    C07D|1976-01-06|
|In an electrostat...|3931469|    H04R|1976-01-06|
|An imaging device...|3931633|    H01L|1976-01-06|
|2-(4-Methoxybenzy...|3932422|    C07D|1976-01-13|
|The compound of c...|3932430|    C07D|1976-01-13|
|Method as claimed...|3932460|    C07D|1976-01-13|
|Apparatus for rap...|3932762|    H01Q|1976-01-13|
|The circuit accor...|3932770|    H03K|1976-01-13|
|A control system ...|3932773| 

Lets look at the distribution of patent ids

In [4]:
df.withColumn("id",col("id").cast(LongType))
    .select("id")
    .summary()
    .show()

+-------+------------------+
|summary|                id|
+-------+------------------+
|  count|           6452014|
|   mean| 7174195.463987524|
| stddev|1874859.4698489187|
|    min|           3930271|
|    25%|           5549093|
|    50%|           7172795|
|    75%|           8795683|
|    max|          10426069|
+-------+------------------+



The distribution of the ids makes sense, Now lets count the number of nulls in the rest of the columns 




In [6]:
df.filter(df("claims") === null).count()

0

Now lets look at the distribution of the sub classes 

In [8]:
val subClasses = df.groupBy("group_id").count().cache()
subClasses.show()

+--------+-----+
|group_id|count|
+--------+-----+
|    D06M| 3850|
|    C05C|  337|
|    C12H|  277|
|    B60T|15636|
|    B63B|12010|
|    A01B| 4917|
|    C23C|24419|
|    D06G|   60|
|    A63K|  125|
|    E06B|11297|
|    E03C| 3538|
|    B27H|   41|
|    A21C| 1940|
|    A24F| 2516|
|    G21G|  409|
|    C10M| 7720|
|    B03C| 2857|
|    F41C| 2424|
|    A46B| 4204|
|    F01P| 2977|
+--------+-----+
only showing top 20 rows



In [9]:
subClasses.select("count").summary().show()

+-------+------------------+
|summary|             count|
+-------+------------------+
|  count|               641|
|   mean| 10065.54446177847|
| stddev|27660.445523882314|
|    min|                 1|
|    25%|               810|
|    50%|              2826|
|    75%|              8815|
|    max|            404420|
+-------+------------------+



In [10]:
val low_count = subClasses.sort("count")
low_count.show(100)

+--------+-----+
|group_id|count|
+--------+-----+
|    Y02B|    1|
|    G06D|    1|
|    A61P|    1|
|    B29K|    1|
|    Y02W|    1|
|    Y02D|    2|
|    Y02E|    3|
|    Y02P|    6|
|    Y02A|    7|
|    B68F|    8|
|    C12L|    9|
|    G21J|   10|
|    C10H|   12|
|    F21W|   13|
|    C06F|   14|
|    F16S|   14|
|    G16Z|   14|
|    E02C|   19|
|    B41G|   20|
|    F21H|   20|
|    F17B|   23|
|    B41D|   26|
|    D01C|   28|
|    B27J|   33|
|    C09H|   35|
|    G10B|   39|
|    B27H|   41|
|    C12J|   41|
|    C12F|   42|
|    D06J|   43|
|    B62C|   48|
|    C10F|   55|
|    D06G|   60|
|    B01B|   69|
|    H05C|   78|
|    B61J|   78|
|    D04G|   81|
|    Y10S|   85|
|    C23D|   91|
|    B31C|   99|
|    F22G|  110|
|    A63K|  125|
|    B21G|  126|
|    B82B|  127|
|    G06C|  130|
|    F23H|  138|
|    F24T|  154|
|    G04D|  155|
|    G09D|  155|
|    G21H|  164|
|    C09F|  167|
|    D01B|  179|
|    D02H|  182|
|    F22D|  188|
|    F24V|  189|
|    G12B|  19

In [11]:
low_count.limit(100).agg(sum($"count")).show()

+----------+
|sum(count)|
+----------+
|     16234|
+----------+



The sub classes are extremely unbalanced. The 100 least used classes are used about 16234 times 


<br>


